### DISCUSSION TITLE
post-processing after get the derivatives of  's' at its sampling points

### DISCUSSION THREAD

**Discussion by oridong** [hungpham2511/toppra#205] [Q&A]
I have two questions about "AVP-biRRT" dealing with the constraints presented in the paper "Dynamic Non-Prehensile Object Transportation".
1) After computation, I get the first time and the second time derivatives of 's' at its sampling points. How could apply this information get 's(t)' as the function of time 't'?
2) When the controller interpolation period is 2ms perhaps usually small compared with the discrete step of 's', this means there are several time-interpolation instants between two consecutive sampling 's', then how to guarantee satisfaction of the constraints in "Dynamic Non-Prehensile Object Transportation" at every time interpolation instant?

**Comment by hungpham2511** [comment]
> After computation, I get the first time and the second time derivatives of 's' at its sampling points. How could apply this information get 's(t)' as the function of time 't'?

There are several ways. These are implemented in the "Parametrizer" classes in both the Python and C++ API.


When the controller interpolation period is 2ms perhaps usually small compared with the discrete step of 's', this means there are several time-interpolation instants between two consecutive sampling 's', then how to guarantee satisfaction of the constraints in "Dynamic Non-Prehensile Object Transportation" at every time interpolation instant?

It's slightly more complex than that. The main idea is since the path and the constraints are "continuous", even if we consider only the constraints at discrete time-point, we can still say that the constraints are satisfied at any points in-between thought discrete points. This idea is similar to algorithms in numerical integration of ODE.

You can check the "InterpolationType" class for more details regarding this.

**Reply by oridong** [reply]
Could you give me some hint of the methods behind  "Parametrizer", I know the matrix form of B-spline and its derivatives. In addition to the constraints of the first time and the second time derivatives of 's' at its sampling points, the arc-lengths between two consecutive sampling 's' are equal and their sum is 1.

**Reply by hungpham2511** [reply]
The paper and the code should be sufficient to describe our method. I am terribly sorry but I don't think I can provide any better explanation.

Do let me know if you have any question on these resources.

**Reply by oridong** [reply]
I mean, how to transform the curve int the (s, sd) space into the (time, s) space, that it, the horizontal axis changes from arc-length to time and the vertical axis changes from the time derivative to the arc-length.

**Reply by hungpham2511** [reply]
One way that this can be done is to assume that the path acceleration is constant (sdd). With this you can calculate everything you want w.r.t time. Hope this helps.
